spec.emu

```diff
@@ -502,54 +1376,148 @@ contributors: Luca Casonato, Guy Bedford, Nicolò Ribaudo
               1. If _state_.[[PendingModulesCount]] = 0, then
                 1. Set _state_.[[IsLoading]] to *false*.
-                1. For each Cyclic Module Record _loaded_ in _state_.[[Visited]], do
+                1. For each Cyclic Module Record _loaded_ of _state_.[[Visited]], do
                   1. If _loaded_.[[Status]] is ~new~, set _loaded_.[[Status]] to ~unlinked~.
-                1. Perform ! Call(_state_.[[PromiseCapability]].[[Resolve]], *undefined*, &laquo; *undefined* &raquo;).
+                1. Perform ! Call(_state_.[[PromiseCapability]].[[Resolve]], *undefined*, « *undefined* »).
               1. Return ~unused~.
             </emu-alg>
           </emu-clause>
 
-          <emu-clause id="sec-ContinueModuleLoading" type="abstract operation">
+          <emu-clause id="sec-InnerModuleLinking" type="abstract operation" oldids="sec-innermoduleinstantiation">
             <h1>
-              ContinueModuleLoading (
-                _state_: a GraphLoadingState Record,
-                <ins>_referrer_: a Cyclic Module Record,</ins>
-                <ins>_specifier_: a String,</ins>
-                _moduleCompletion_: either a normal completion containing a Module Record or a throw completion,
-              ): ~unused~
+              InnerModuleLinking (
+                _module_: a Module Record,
+                _stack_: a List of Cyclic Module Records,
+                _index_: a non-negative integer,
+              ): either a normal completion containing a non-negative integer or a throw completion
+            </h1>
+            <dl class="header">
+              <dt>description</dt>
+              <dd>It is used by Link to perform the actual linking process for _module_, as well as recursively on all other modules in the dependency graph. The _stack_ and _index_ parameters, as well as a module's [[DFSIndex]] and [[DFSAncestorIndex]] fields, keep track of the depth-first search (DFS) traversal. In particular, [[DFSAncestorIndex]] is used to discover strongly connected components (SCCs), such that all modules in an SCC transition to ~linked~ together.</dd>
+            </dl>
+
+            <emu-alg>
+              1. If _module_ is not a Cyclic Module Record, then
+                1. Perform ? _module_.Link().
+                1. Return _index_.
+              1. If _module_.[[Status]] is one of ~linking~, ~linked~, ~evaluating-async~, or ~evaluated~, then
+                1. Return _index_.
+              1. Assert: _module_.[[Status]] is ~unlinked~.
+              1. Set _module_.[[Status]] to ~linking~.
+              1. Set _module_.[[DFSIndex]] to _index_.
+              1. Set _module_.[[DFSAncestorIndex]] to _index_.
+              1. Set _index_ to _index_ + 1.
+              1. Append _module_ to _stack_.
+              1. For each <del>String</del><ins>ModuleRequest</ins> _required_ of _module_.[[RequestedModules]], do
+                1. <ins>If _required_.[[Phase]] is ~evaluation~, then</ins>
```

Comment:
We should fail linking if the source is not available, right?

```
1. Else,
  1. Assert: _required_.[[Phase]] is ~source~.
  1. If _requiredModule_.[[ModuleSourceObject]] is *null*, throw a SyntaxError.
```

(SyntaxError feels wrong, but it's what we use for other linking errors)

Reply:
Thanks, I've made this a `ReferenceError` for now.